packages/block-editor/src/components/rich-text/index.js

```diff
@@ -354,7 +354,7 @@ function RichTextWrapper(
 			{ isSelected && hasFormats && (
 				<FormatToolbarContainer
 					inline={ inlineToolbar }
-					anchorRef={ anchorRef }
+					editableContentElement={ anchorRef.current }
```

Comment:
> Refactoring `anchorRef` to anchor is the simplest and most frequent case — it's mostly about passing the actual element instead of a reference.

I think we need to be careful of this case when refactoring. Reading refs during render is [not recommended](https://github.com/facebook/react/pull/18545) and may throw a warning in the future. I think that's also why `floating-ui` recommends `useState` for storing the element instance rather than `useRef`.

I wonder if there's a way to warn against usage like this on the API level 🤔?

Reply:
This is tricky.

References were already used prior to this PR (for this specific case, they were read _inside_ the `FormatToolbarContainer` component as part of the `useAnchorRef` hook, and my refactor just moved the reading outside of the component), and therefore I don't think that this PR really introduces this patter.

If anything, this PR slightly improves the situation by deprecating the `anchorRef` prop, which was previously accepting `ref`s instead of actual values.

> I think that's also why `floating-ui` recommends `useState` for storing the element instance rather than `useRef`.

That's correct, and in fact I tried to refactor as many parts of the code as possible to `useState` — I just didn't apply the refactor around the `rich-text` components because these references are forwarded from outside of these components, and I didn't want to grow the scope of this PR.

> I wonder if there's a way to warn against usage like this on the API level 🤔?

I don't think so, at least from the `Popover`'s point of view, since the component doesn't have a way of knowing if the value passed as the `anchor` comes from a `ref` or from internal state.

I did add a few paragraphs in the component's README about using internal state in order to ensure reactive updates (https://github.com/WordPress/gutenberg/pull/43691/commits/74af7eb5a16d9ede795533b25429fa3fddee62a0)

Reply:
Yeah, it's definitely tricky 🤔, I don't have a solution yet too.

I think `ref`s are fine though, as long as we don't read them during render. A solution to that would be using a function to get the reference element (a separate `getAnchor` prop or extend `anchor`).

```diff
- anchor={ ref.current }
+ anchor={ () => ref.current }
```

As long as the underlying implementation of `<Popover>` doesn't call the function during render then we'll be fine.

But then it all just feels like an attempt to fix a false positive though, I'm not sure if it's worth it 😅. Let's keep it as is until we find a better solution?

Reply:
> But then it all just feels like an attempt to fix a false positive though, I'm not sure if it's worth it 😅. Let's keep it as is until we find a better solution?

Agreed. The best way to solve this problem would be to refactor the code outside of the `Popover` and use internal state instead (as I already did partially in this PR), so maybe we can work on that refactor at a later point if/when necessary.